cocotb/triggers.py

```diff
@@ -731,6 +809,70 @@ def _wait(self):
         raise ReturnValue(self)
 
 
+class WaitCondition(Waitable):
+    """
+    Wait until a given condition is True state.
+
+    Normally this class won't be used directly but through :meth:`TriggerHelper.until`,
+    :meth:`TriggerHelper.until_after`, :meth:`TriggerHelper.do_until` and
+    :meth:`TriggerHelper.do_until_after`
+
+    Args:
+        condition (callable): Output of ``condition()`` is checked to be equivalent to logic
+            ``True`` for condition to be met.
+        event (Trigger, optional): event after which to check the condition; defaults to
+            :class:`NextTimeStep`.
+        readonly (bool, optional): whether to do check in :class:`ReadOnly` state;
+            defaults to ``True``.
+        before (bool, optional): whether to do the event once before doing first check;
+            defaults to ``False``.
+        after (bool, optional): whether to wait for an extra event after condition
+            is met; defaults to ``False``.
+
+    Examples:
+
+        Wait for ack on a synchronous bus
+
+        .. code-block:: python
+
+            yield WaitCondition(
+                lambda: bus.cycle.integer and bus.ack.integer,
+                event=RisingEdge(dut.clk)
+            )
+
+        Wait for asynchronous ack
+
+        .. code-block:: python
+
+            yield WaitCondition(lambda: bus.ack.integer, Edge(bus.ack))
+
+    .. versionadded: 1.3
+    """
+    def __init__(self, condition, **kwargs):
+        self.condition = condition
+        self.event = kwargs.pop('event', NextTimeStep())
```

Comment:
I think it's worth a type check here that `event` is either a `Trigger` or `Waitable`, this will behave poorly if a coroutine is passed.

Reply:
OK